### PR TITLE
Feature/user schema

### PR DESCRIPTION
**Pull request by gibsonhan**
Requesting Pull: Request. 

I humbly ask for a code review. I am not sure what to do next.
I have created the UserScheme and use mongoose-validator, and custom-mongoose-validator to validate dator. I have not gotten to the , mutator, or resolver. In fact I m not sure where to begin.

### CODE REVIEW

**Review by charlesjliu — CHANGES_REQUESTED** [review]
@gibsonhan This looks like a solid start. I'd recommend writing some Jest tests to test creating users.

**Comment by gibsonhan** [comment]
@charlesjliu will do. Btw is this proper etiquette to confirm and reply to your request changes on github?

File: src/schema/user.graphql [hunk 1]
```diff
@@ -1,20 +1,78 @@
...
-  id: ID
-  name: String
-  type: UserType
+    id: [ID!]
+    first: String
+    last: String
+    email: [String!]
+    appearsIn: [social_type]!
+    password: [String!]
+    failed_login: {
```
Comment (charlesjliu) [inline src/schema/user.graphql]: This won't work as valid GraphQL schema syntax, here you want to define the type for `User`. Think about it more as, what fields do would we want to return to the UI? There doesn't need to be 1:1 mapping of every field in the Mongo schema to the `User` type  we define in the GraphQL schema.

File: src/schema/user.graphql [hunk 2]
```diff
@@ -1,20 +1,78 @@
...
+input password {
+    password: String!
 }
 
 type User {
-  id: ID
-  name: String
-  type: UserType
+    id: [ID!]
+    first: String
```
Comment (charlesjliu) [inline src/schema/user.graphql]: What do you think about getting rid of `first` and `last` and replacing it with just `name`?

File: src/models/user.js [hunk 3]
```diff
@@ -1,9 +1,136 @@
...
+
+  validate({
+    valiator: str => isValidPassword(str, { minlength: 10 }),
+    message:
+      'Password must have at least: 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character.',
+  }),
+];
+
+const UserSchema = new Schema({
+  name: {
```
Comment (charlesjliu) [inline src/models/user.js]: Thoughts on just making `first_name` and `last_name` top-level fields and not nested under `name`? seems unnecessary to nest them. We can add `name` as a virtual property to the schema instead.

We should also use camelCase for field names 😄

File: src/models/user.js [hunk 4]
```diff
@@ -1,9 +1,136 @@
...
+    type: String,
+    require: true,
+    unique: true,
+    lowercase: true,
+    trim: true,
+    allowNull: false,
+    validate: emailValidator,
+  },
+
+  social_media: {
```
Comment (charlesjliu) [inline src/models/user.js]: camel case!

File: src/models/user.js [hunk 5]
```diff
@@ -1,9 +1,136 @@
-import mongoose from 'mongoose';
+import { mongoose } from 'mongoose';
+import { bcrypt } from 'bcrypt-nodejs';
+import { validate } from 'mongoose-validator';
+import { isValidPassword } from 'mongoose-custom-validators';
 
-const UserSchema = new mongoose.Schema({
-  name: 'string',
-  type: 'string',
+const { SALT_INDEX } = 10;
```
Comment (charlesjliu) [inline src/models/user.js]: should this be `const SALT_INDEX = 10` ?

File: src/schema/user.graphql [hunk 6]
```diff
@@ -1,20 +1,78 @@
-enum UserType {
-  Mentor
-  Mentee
-  Both
+enum social_type {
```
Comment (charlesjliu) [inline src/schema/user.graphql]: Use Pascal case for all GraphQL schema type definitions

**Review by charlesjliu — CHANGES_REQUESTED** [review]
@gibsonhan okay I looked at it a little more. Left some comments but I later realized this might be your first time working with GraphQL (which is great!) so let's tackle this iteratively. I'll DM you on slack.

**Review by charlesjliu — APPROVED** [review]

File: src/models/user.js [hunk 7]
```diff
@@ -1,9 +1,134 @@
...
+    message: 'There can be no +, -, or .',
+  }),
+];
+
+const emailValidator = [
+  validate({
+    type: String,
+    valiator: 'isLength',
+    arguments: [3, 32],
+    message: 'Name should be btween {ARGS[0]} and {ARGS[1]} characters',
```
Comment (jagzviruz) [inline src/models/user.js]: Please correct the message text.

File: src/models/user.js [hunk 8]
```diff
@@ -1,9 +1,134 @@
...
+    message:
+      'Password must have at least: 1 uppercase letter, 1 lowercase letter, 1 number, and 1 special character.',
+  }),
+];
+
+const userSchema = new Schema({
+  firstName: {
+    type: String,
+    require: true,
+    lowercase: true,
```
Comment (jagzviruz) [inline src/models/user.js]: It would be better to leave the case untouched. Let the user control the case of his name.

File: src/models/user.js [hunk 9]
```diff
@@ -1,9 +1,134 @@
...
+    trim: true,
+    allowNull: false,
+    validate: nameValidator,
+  },
+
+  lastName: {
+    type: String,
+    require: true,
+    lowercase: true,
```
Comment (jagzviruz) [inline src/models/user.js]: It would be better to leave the case untouched. Let the user control the case of his name.

File: src/models/user.js [hunk 10]
```diff
@@ -1,9 +1,134 @@
...
+    },
+  },
+
+  password: {
+    type: String,
+    require: true,
+    minlength: 8,
+    maxlength: 32,
+    trim: true,
+    allowNull: false,
```
Comment (jagzviruz) [inline src/models/user.js]: If a user uses social login, he should not have to enter a password.

File: src/models/user.js [hunk 11]
```diff
@@ -1,9 +1,134 @@
...
+    failedLogin: {
+      lastAttempt: Date,
+      numFailed: {
+        type: Number,
+        default: 0,
+        max: 5,
+      },
+    },
+  },
 });
```
Comment (jagzviruz) [inline src/models/user.js:130]: We might need a role and activation status for the user. Please add sensible defaults for now.

File: src/models/user.js (hunk 10 above)
Comment (gibsonhan) [reply src/models/user.js]: should I add a validator to or some kind of check to test if the social media is link or do I just remove the AllowNull argument?

File: src/models/user.js (hunk 11 above)
Comment (gibsonhan) [reply src/models/user.js:130]: "sensible defaut" can you clarify what this means 

image

or do you mean

this "[sensible default"](https://stevebennett.co/2017/07/24/the-power-of-sensible-defaults/)

**Pull request merged**